File: src/compiler.ts
import type { Program } from "./ast.js";
import { resolveBinding, type BindingSpec } from "./binding.js";
import { checkAAL } from "./checker.js";
import type { Diagnostic } from "./diagnostics.js";
import { generateTypeScript } from "./codegen.js";
import { parseAAL } from "./parser.js";
import { DEFAULT_LANGUAGE, type AALLanguage } from "./language.js";

export interface CompileResult {
  readonly program: Program | null;
  readonly diagnostics: readonly Diagnostic[];
  readonly code: string | null;
}

export interface CompileOptions {
  readonly binding?: BindingSpec;
  readonly language?: AALLanguage;
}

export function compileAAL(source: string, options: CompileOptions = {}): CompileResult {
  const language = options.language ?? DEFAULT_LANGUAGE;
  const parsed = parseAAL(source, { language });
  if (!parsed.program || parsed.diagnostics.some((diagnostic) => diagnostic.severity === "error")) {
    return { program: parsed.program, diagnostics: parsed.diagnostics, code: null };
  }

  const semanticDiagnostics = checkAAL(parsed.program, language);
  if (semanticDiagnostics.some((diagnostic) => diagnostic.severity === "error")) {
    return { program: parsed.program, diagnostics: semanticDiagnostics, code: null };
  }

  const bindingResult = resolveBinding(parsed.program, options.binding, language);
  const diagnostics = [...semanticDiagnostics, ...bindingResult.diagnostics];
  if (!bindingResult.binding || diagnostics.some((diagnostic) => diagnostic.severity === "error")) {
    return { program: parsed.program, diagnostics, code: null };
  }

  return { program: parsed.program, diagnostics, code: generateTypeScript(parsed.program, bindingResult.binding, language) };
}

File: internal/compiler/compiler.go
package compiler

import (
	"github.com/myczh-1/determinant/internal/ast"
	"github.com/myczh-1/determinant/internal/binding"
	"github.com/myczh-1/determinant/internal/diagnostics"
	"github.com/myczh-1/determinant/internal/language"
	"github.com/myczh-1/determinant/internal/model"
	"github.com/myczh-1/determinant/internal/parser"
	"github.com/myczh-1/determinant/internal/semantic"
)

type Result struct {
	Program     *ast.Program
	TypeInfo    *semantic.TypeInfo
	Binding     *binding.Resolved
	Model       model.Program
	Diagnostics []diagnostics.Diagnostic
}

func Compile(source string, lang language.Language, file string) Result {
	parsed := parser.Parse(source, lang, file)
	result := Result{Program: parsed.Program, Diagnostics: append([]diagnostics.Diagnostic(nil), parsed.Diagnostics...)}
	if parsed.Program == nil || hasErrors(result.Diagnostics) {
		return result
	}
	typeInfo, semanticDiagnostics := semantic.Check(parsed.Program)
	result.TypeInfo = typeInfo
	result.Diagnostics = append(result.Diagnostics, semanticDiagnostics...)
	if !hasErrors(result.Diagnostics) {
		result.Model = model.FromAST(parsed.Program, typeInfo)
	}
	return result
}

func hasErrors(items []diagnostics.Diagnostic) bool {
	for _, item := range items {
		if item.Severity == diagnostics.Error {
			return true
		}
	}
	return false
}
